modules/core/src/smithy4s/http/internals/UrlFormDataDecoderSchemaVisitor.scala

```diff
@@ -0,0 +1,225 @@
+                    _
+                  ) =>
+                index -> formData
+            }
+            .groupBy { case (index, _) =>
+              index
+            }
+            .toVector
+            .sortBy { case (index, _) =>
+              index
+            }
+            .map { case (index, indicesAndValues) =>
+              UrlFormCursor(
+                history,
+                indicesAndValues.map { case (_, value) => value }
+              )
+                .down(PayloadPath.Segment.Index(index))
+            }
+          groupedAndSortedCursors
+            .traverse[UrlFormDecodeError, A](memberDecoder.decode(_))
+            .map(list => tag.fromIterator(list.iterator))
+      }
+  }
+
+  override def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): UrlFormDataDecoder[Map[K, V]] = {
+    type KV = (K, V)
+    val kvSchema: Schema[(K, V)] = {
+      val kField = key.required[KV]("key", _._1)
+      val vField = value.required[KV]("value", _._2)
+      Schema.struct(kField, vField)((_, _)).addHints(UrlFormName("entry"))
+    }
+    compile(Schema.vector(kvSchema).addHints(hints))
+      .map(_.toMap)
+  }
+
+  override def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: EnumTag[E],
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): UrlFormDataDecoder[E] = tag match {
+    case EnumTag.IntEnum() =>
+      val desc = s"enum[${values.map(_.intValue).mkString(", ")}]"
+      val valueMap = values.map(ev => ev.intValue -> ev.value).toMap
+      UrlFormDataDecoder.fromStringParser(desc)(
+        _.toIntOption.flatMap(valueMap.get(_))
+      )
+
+    case _ =>
+      val desc = s"enum[${values.map(_.stringValue).mkString(", ")}]"
+      val valueMap = values.map(ev => ev.stringValue -> ev.value).toMap
+      UrlFormDataDecoder.fromStringParser(desc)(valueMap.get)
+  }
+
+  override def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[Field[S, _]],
+      make: IndexedSeq[Any] => S
+  ): UrlFormDataDecoder[S] = {
+    def fieldDecoder[A](field: Field[S, A]): UrlFormDataDecoder[A] =
+      compile(field.schema).down(getKey(field.hints, field.label))
+    val decoders = fields.map(fieldDecoder(_))
+    cursor =>
+      decoders
+        .traverse((decoder: UrlFormDataDecoder[_]) => decoder.decode(cursor))
+        .map(make)
+  }
+
+  override def union[U](
+      shapeId: ShapeId,
+      hints: Hints,
+      alternatives: Vector[Alt[U, _]],
+      dispatch: Alt.Dispatcher[U]
+  ): UrlFormDataDecoder[U] = {
+    def altDecoder[A](
+        alt: Alt[U, A]
+    ): (PayloadPath.Segment, UrlFormDataDecoder[U]) = {
+      val key = getKey(alt.hints, alt.label)
+      val decoder = compile(alt.schema).map(alt.inject).down(key)
+      (key, decoder)
+    }
+    val altMap = alternatives
+      .map(altDecoder(_))
+      .toMap[PayloadPath.Segment, UrlFormDataDecoder[U]]
+    ({
```

Comment:
`locally` is your friend. The parenthesis made me think this was a function application